src/main/scala/mesosphere/marathon/api/AuthResource.scala

```diff
@@ -1,52 +1,67 @@
 package mesosphere.marathon.api
 
-import javax.servlet.http.{ HttpServletRequest, HttpServletResponse }
+import javax.servlet.http.HttpServletRequest
 import javax.ws.rs.core.Response
 
+import mesosphere.marathon.{ UnknownGroupException, UnknownAppException, AccessDeniedException }
 import mesosphere.marathon.plugin.auth._
 import mesosphere.marathon.plugin.http.HttpResponse
-import mesosphere.marathon.state.PathId
 
 /**
   * Base trait for authentication and authorization in http resource endpoints.
   */
 trait AuthResource extends RestResource {
-
   def authenticator: Authenticator
   def authorizer: Authorizer
 
-  def doIfAuthenticated(request: HttpServletRequest,
-                        response: HttpServletResponse)(fn: Identity => Response): Response = {
+  def authenticated(request: HttpServletRequest)(fn: Identity => Response): Response = {
     val requestWrapper = new RequestFacade(request)
-    val identity = result(authenticator.authenticate(requestWrapper))
-    identity.map(fn).getOrElse {
-      val responseWrapper = new ResponseFacade
-      authenticator.handleNotAuthenticated(requestWrapper, responseWrapper)
-      responseWrapper.response
+    val maybeIdentity = result(authenticator.authenticate(requestWrapper))
+    maybeIdentity.map { identity =>
+      try {
+        fn(identity)
+      }
+      catch {
+        case e: AccessDeniedException => withResponseFacade(authorizer.handleNotAuthorized(identity, _))
+        case e: UnknownAppException   => unknownApp(e.id, e.version)
```

Comment:
Do you have to handle those exceptions here?
The `authenticated` method should definitively handle the AccessDeniedException. I am not sure about the other ones.